log-manager/src/main/java/io/airlift/log/BufferedHandler.java

```diff
@@ -368,6 +419,27 @@ public void flush()
         flushMessageOutputSafe();
     }
 
+    /**
+     * Requests a full flush of all currently queued messages from the background thread. If the handler has already been closed,
+     * this method will return an empty {@link Optional}. Otherwise, the returned {@link Optional} will contain a {@link ListenableFuture}
+     * that will complete when the background flush has completed.
+     */
+    public Optional<ListenableFuture<Void>> requestFullFlush()
+    {
+        queueDrainLock.lock();
+        try {
+            if (inputClosed.get() || terminalMessageDequeued) {
+                return Optional.empty();
+            }
+            SettableFuture<Void> flushedSignal = SettableFuture.create();
+            queueInsert(new MessageAndSource(EMPTY_BYTES, "", flushedSignal));
+            return Optional.of(flushedSignal);
```

Comment:
this should be wrapped in `Futures.nonCancellationPropagating(flushedSignal)` to prevent a caller from disrupting a future shared by another caller (since they are linked with `setFuture`)